Lua/creature adjustments

lahm86 opened the pull request:
Part of #3873.

#### Checklist

- [x] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [x] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description

This resolves a problem with statistics if a level is set to have hostile allies from the get-go (via `on_level_load`). The order of events on develop would be:

- Unload level
  - Set allies hostile = false
- Load level
  - Calculate stats
  - Allies are not hostile, so are not included as kills
- Run the level's Lua script
  - Allies are now hostile; the stats are inaccurate

I'm not convinced introducing another event is the best approach but wanted to get your thoughts.

This can be tested in Barkhang if you add the following to a Lua file and link it in the GF.

```lua
trx.events.on_level_init(function(level)
  trx.creatures.hostile_allies = true
end)
```

This also prepares for allowing allies to be defined externally, so some checks around the kill command, shooting monks, and seeing that they still pick up items would be good (e.g. item 79 in Kingdom).

lahm86 committed:
creature/behavior: introduce basic creature behavior module

lahm86 committed:
creature/behavior: allow treating monks as enemies

This allows monks to be treated as enemies from the get-go in a level,
in terms of ensuring the kill statistics are accurate.

lahm86 committed:
lua: introduce level initialisation event

This allows implementing Lua logic prior to a level loading but after
the previous level  having been unloaded. This for example allows
allies to become aggressive from the get-go and their kills included
in the stats.

lahm86 committed:
docs: update TR2 game flow schema

rr- commented on src/trx/game/carrier.c:
```diff
@@ -54,7 +54,6 @@ static bool M_IsCarrierType(const OBJECT_ID obj_id)
     // also creates issues with /kill, hence a separate check is required here.
     is_enemy &= obj_id != O_EEL && obj_id != O_BIG_EEL;
     is_enemy &= obj_id != O_SKIDOO_DRIVER;
-    is_enemy |= Object_IsType(obj_id, g_AllyObjects);
```
Doesn't this break GM allied monk's drops?

lahm86 replied:
They're part of the enemy array as standard (which is checked first) so this becomes redundant. Although admittedly it'd be harmless to leave it in place.

rr- replied:
It's pretty odd that enemies are a superset of allies. Sounds like an either-or. Wdyt?

lahm86 replied:
It sits ok with me tbh, I'm thinking the ally list is more of an exclusion set rather than a subset. Perhaps we could rename the enemy list to g_CreatureObjects - so conceptually a creature will by default be hostile unless in the ally list?
This should hopefully mean we only need to offer editing the ally and ally-targeting lists in Lua, rather than these plus the global enemy list.

rr- replied:
Yeah I'm in for a rename. Thanks!

lahm86 replied:
Done ☺️

rr- commented on src/trx/game/console/cmd/kill.c:
```diff
@@ -21,7 +21,6 @@ extern OBJECT_ID CombatEnd_GetBossType(void);
 static bool M_CanTargetObjectCreature(const OBJECT_ID obj_id)
 {
     return (Object_IsType(obj_id, g_EnemyObjects)
-            || Object_IsType(obj_id, g_AllyObjects)
```
Is this intended?

lahm86 replied:
Similar to above. Could leave these in place for clarity perhaps?

lahm86 committed:
objects/vars: rename g_EnemyObjects to g_CreatureObjects

This renames the global list of enemies to imply the concept that
any creature is an enemy unless it is defined also as an ally.

rr- reviewed (approved):
Thanks!
